pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -275,6 +323,95 @@ func (r *ReconcileCSINodeTopology) Reconcile(ctx context.Context, request reconc
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileCSINodeTopology) reconcileForGuest(ctx context.Context, request reconcile.Request) (
+	reconcile.Result, error) {
+	log := logger.GetLogger(ctx)
+	log.Infof("Start reconciling the CSINodeTopology request %s in %s", request.Name, cnstypes.CnsClusterFlavorGuest)
+
+	// Fetch the CSINodeTopology instance.
+	instance := &csinodetopologyv1alpha1.CSINodeTopology{}
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("CSINodeTopology resource with name %q not found. Ignoring since object must have "+
+				"been deleted.", request.Name)
+			return reconcile.Result{}, nil
+		}
+		log.Errorf("Failed to fetch the CSINodeTopology instance with name: %q. Error: %+v", request.Name, err)
+		// Error reading the object - return with err.
+		return reconcile.Result{}, err
+	}
+
+	// Initialize backOffDuration for the instance, if required.
+	var timeout time.Duration
+	func() {
+		backOffDurationMapMutex.Lock()
+		defer backOffDurationMapMutex.Unlock()
+		if _, exists := backOffDuration[instance.Name]; !exists {
+			backOffDuration[instance.Name] = time.Second
+		}
+		timeout = backOffDuration[instance.Name]
+	}()
+
+	// Fetch topology labels for guest worker node backed by vmop VM.
+	topologyLabels, err := getNodeTopologyInfoForGuest(ctx, instance, r.vmOperatorClient, r.supervisorNamespace)
+	if err != nil {
+		msg := fmt.Sprintf("failed to fetch topology information for the worker node %q. Error: %v",
+			instance.Name, err)
+		log.Error(msg)
+		_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+
+	// Update CSINodeTopology instance.
+	instance.Status.TopologyLabels = topologyLabels
+	if err := updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologySuccess,
+		fmt.Sprintf("Topology labels successfully updated for the worker node %q", instance.Name)); err != nil {
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+
+	// On successful event, remove instance from backOffDuration.
+	func() {
+		backOffDurationMapMutex.Lock()
+		defer backOffDurationMapMutex.Unlock()
+		delete(backOffDuration, instance.Name)
+	}()
+
+	log.Infof("Successfully updated topology labels for worker %q in %s",
+		instance.Name, cnstypes.CnsClusterFlavorGuest)
+	return reconcile.Result{}, nil
+}
+
+func getNodeTopologyInfoForGuest(ctx context.Context, instance *csinodetopologyv1alpha1.CSINodeTopology,
+	vmOperatorClient client.Client, supervisorNamespace string) ([]csinodetopologyv1alpha1.TopologyLabel, error) {
+	log := logger.GetLogger(ctx)
+
+	virtualMachine := &vmoperatortypes.VirtualMachine{}
+	vmKey := types.NamespacedName{
+		Namespace: supervisorNamespace,
+		Name:      instance.Name, // use the nodeName as the VM key
```

Comment:
No, it is not safe. `Spec.NodeID` could come from system UUID if UseCSINodeId feature is enabled (even though this feature doesn't plan to support TKGs, it might be in future.), in which case Spec.NodeID is different from VMOP VM name. `instance.Name` coming from [nodeName](https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/99d8f9c9a33f5d6406b7d562da6b0262d8367073/manifests/guestcluster/1.22/pvcsi.yaml#L374) would be more reliable.